core/src/main/java/io/grpc/internal/OobChannel.java

```diff
@@ -0,0 +1,244 @@
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.LoadBalancer2.PickResult;
+import io.grpc.LoadBalancer2.SubchannelPicker;
+import io.grpc.ManagedChannel;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import io.grpc.internal.ClientCallImpl.ClientTransportProvider;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A ManagedChannel backed by a single {@link InternalSubchannel} and used for {@link LoadBalancer2}
+ * to its own RPC needs.
+ */
+@ThreadSafe
+final class OobChannel extends ManagedChannel implements WithLogId {
+  private static final Logger log = Logger.getLogger(OobChannel.class.getName());
+
+  private SubchannelImpl subchannelImpl;
+  private SubchannelPicker subchannelPicker;
+
+  private final LogId logId = LogId.allocate(getClass().getName());
+  private final CensusContextFactory censusFactory;
+  private final String authority;
+  private final DelayedClientTransport2 delayedTransport;
+  private final Executor executor;
+  private final ScheduledExecutorService deadlineCancellationExecutor;
+  private final Supplier<Stopwatch> stopwatchSupplier;
+  private final CountDownLatch terminatedLatch = new CountDownLatch(1);
+  private volatile boolean shutdown;
+
+  private final ClientTransportProvider transportProvider = new ClientTransportProvider() {
+    @Override
+    public ClientTransport get(CallOptions callOptions, Metadata headers) {
+      // delayed transport's newStream() always acquires a lock, but concurrent performance doesn't
+      // matter here because OOB communication should be sparse, and it's not on application RPC's
+      // critical path.
+      return delayedTransport;
+    }
+  };
+
+  OobChannel(CensusContextFactory censusFactory, String authority, Executor executor,
+      ScheduledExecutorService deadlineCancellationExecutor, Supplier<Stopwatch> stopwatchSupplier,
+      ChannelExecutor channelExecutor) {
+    this.censusFactory = checkNotNull(censusFactory, "censusFactory");
+    this.authority = checkNotNull(authority, "authority");
+    this.executor = checkNotNull(executor, "executor");
+    this.deadlineCancellationExecutor = checkNotNull(
+        deadlineCancellationExecutor, "deadlineCancellationExecutor");
+    this.stopwatchSupplier = checkNotNull(stopwatchSupplier, "stopwatchSupplier");
+    this.delayedTransport = new DelayedClientTransport2(executor, channelExecutor);
+    this.delayedTransport.start(new ManagedClientTransport.Listener() {
+        @Override
+        public void transportShutdown(Status s) {
+          // Don't care
+        }
+
+        @Override
+        public void transportTerminated() {
+          subchannelImpl.shutdown();
```

Comment:
(still learning the code..) so we want to shut down the sub-channel when an oob-channel is terminated, not when the transport is shut down?